// react-form-with-constraints/lib-es5/Field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FieldFeedbackType_1 = require("./FieldFeedbackType");
var clearArray_1 = require("./clearArray");
var Field = (function () {
    function Field(name) {
        this.name = name;
        this.validations = [];
    }
    Field.prototype.addOrReplaceValidation = function (validation) {
        var i = this.validations.findIndex(function (_validation) { return _validation.key === validation.key; });
        if (i > -1)
            this.validations[i] = validation;
        else
            this.validations.push(validation);
    };
    Field.prototype.clearValidations = function () {
        clearArray_1.default(this.validations);
    };
    Field.prototype.hasFeedbacksOfType = function (type, fieldFeedbacksKey) {
        return this.validations.some(function (fieldFeedback) {
            return (fieldFeedbacksKey === undefined || fieldFeedback.key.startsWith(fieldFeedbacksKey + ".")) &&
                fieldFeedback.type === type && fieldFeedback.show === true;
        });
    };
    Field.prototype.hasErrors = function (fieldFeedbacksKey) {
        return this.hasFeedbacksOfType(FieldFeedbackType_1.default.Error, fieldFeedbacksKey);
    };
    Field.prototype.hasWarnings = function (fieldFeedbacksKey) {
        return this.hasFeedbacksOfType(FieldFeedbackType_1.default.Warning, fieldFeedbacksKey);
    };
    Field.prototype.hasInfos = function (fieldFeedbacksKey) {
        return this.hasFeedbacksOfType(FieldFeedbackType_1.default.Info, fieldFeedbacksKey);
    };
    Field.prototype.hasFeedbacks = function (fieldFeedbacksKey) {
        return this.hasErrors(fieldFeedbacksKey) ||
            this.hasWarnings(fieldFeedbacksKey) ||
            this.hasInfos(fieldFeedbacksKey);
    };
    Field.prototype.isValid = function () {
        return !this.hasErrors();
    };
    return Field;
}());
exports.default = Field;
