// react-form-with-constraints/lib-es5/InputElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var InputElement = (function () {
    function InputElement(input) {
        if (input.props === undefined) {
            input = input;
            this.name = input.name;
            this.type = input.type;
            this.value = input.value;
            this.validity = new IValidityState(input.validity);
            this.validationMessage = input.validationMessage;
        }
        else {
            input = input;
            this.name = input.props.name;
            this.type = undefined;
            this.value = input.props.value;
            this.validity = undefined;
            this.validationMessage = undefined;
        }
    }
    return InputElement;
}());
exports.InputElement = InputElement;
var IValidityState = (function () {
    function IValidityState(validity) {
        this.badInput = validity.badInput;
        this.customError = validity.customError;
        this.patternMismatch = validity.patternMismatch;
        this.rangeOverflow = validity.rangeOverflow;
        this.rangeUnderflow = validity.rangeUnderflow;
        this.stepMismatch = validity.stepMismatch;
        this.tooLong = validity.tooLong;
        this.tooShort = validity.tooShort;
        this.typeMismatch = validity.typeMismatch;
        this.valid = validity.valid;
        this.valueMissing = validity.valueMissing;
    }
    return IValidityState;
}());
exports.IValidityState = IValidityState;
